Dozuki/CloudPrem-Infra
# BI Database VPN Access & VPN Module

# Background
Customers will likely need access to their BI database from outside the AWS network. Our existing configuration relies on their account having an existing transit gateway managed outside the terraform. 

# Solution
We will now conditionally provision a VPN endpoint to the VPC which will allow access to authenticated users. The VPN configuration is such that only the BI database will be accessible from within the VPN and the user may configure which CIDRs they want to allow to connect to the VPN. Our VPN uses mutual authentication which relies on 2048 bit RSA self-signed TLS certificates. We generate and upload the certificates embedded in an OpenVPN configuration file to a secure S3 bucket for distribution. The bucket name and BI Database secret containing the connection information is output after the successful application of the physical module.

# QA
To test this change, download the `bi-vpn` branch, navigate to a `bi` testing environment folder in your account and region of choice, be sure the `bi_vpn_access` variable is set to `true` and the `bi_access_cidrs` variable is set to a cidr you can connect from and then apply the stack as usual. 

Once application is complete you can query the outputs for the physical module, grab the `bi_database_credential_secret` and `bi_vpn_configuration_bucket` values. The secret contains the mysql endpoint and credentials for the bi database and the configuration bucket contains the OpenVPN configuration file. Download that file to your workstation and load it into your VPN program of choice and connect. Once successfully connected you should be able to make a MySQL connection to the BI database using the credentials in the secret.

# Notable Changes
Along with the main VPN additions there are some other minor fixes as well including the following:

1. Complete S3 bucket names
   * We no longer use generic bucket prefixes, we will use the identifier and region to name the buckets something that can be easily found in the S3 dashboard
2. DMS stop script race condition mitigation
   * There seems to be some lag in the API between when a DMS task is stopped vs when it's actually reported as stopped in subsequent API calls so I added a 5 minute sleep to the end of the script to give the API time to update. This seems to have fixed this issue in my tests.
3. Bastion broken Curl & Helm Install
   * kubectl install has been broken so I fixed that. I also added helm install since that is a helpful utility to have available for debugging.
4. Tighter security for databases
   * To facilitate secure VPN usage I tightened up the security groups for the databases. We now only allow access to what's necessary instead of just allowing the entire VPC.
